mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/impl/PojoMassIndexingIndexedTypeGroup.java

```diff
@@ -48,16 +48,18 @@ public static List<PojoMassIndexingIndexedTypeGroup<?>> disjoint(PojoMassIndexin
 		for ( PojoMassIndexingIndexedTypeContext<?> typeContext : indexedTypeContexts ) {
 			PojoMassIndexingIndexedTypeGroup<?> typeGroup = PojoMassIndexingIndexedTypeGroup.single(
 					mappingContext, typeContextProvider, typeContext, loadingContext );
-			// First try to merge this new type group with an existing one
-			ListIterator<PojoMassIndexingIndexedTypeGroup<?>> iterator = typeGroups.listIterator();
-			while ( iterator.hasNext() ) {
-				PojoMassIndexingIndexedTypeGroup<?> mergeResult = iterator.next().mergeOrNull( typeGroup );
-				if ( mergeResult != null ) {
-					// We found an existing group that can be merged with this one.
-					// Remove that group, we'll add the merge result to the list later.
-					typeGroup = mergeResult;
-					iterator.remove();
-					// Continue iterating through existing groups, as we may be able to merge with multiple groups.
+			if ( !typeContext.canPotentiallyBeJoined() ) {
```

Comment:
Are you sure about this implementation? Believe it or not, when I wrote "cancelling grouping for JOINED entities _when not all concrete entities are present in the group_", the "not all concrete entities" part was there on purpose :grin: 

Your current solution could actually result in worse results when e.g. we have classes `A`, `B extends A`, and `C extends A`, all of them concrete (non-abstract), and we try to reindex A (plus subclasses): we'd end up issuing one query for B (fine), one for C (fine), and one for A (not so fine, would be something like `A join B join C where B.id is null and C.id is null`). All that when a single query targeting all these types would have worked just fine...

Critically, what I mentioned just now is probably much more frequent than the problem you're effectively solving here, which is when we try to reindex A (which is indexed), but C is not indexed.

I'd actually expect this:

* TABLE_PER_CLASS => no grouping
* SINGLE_TABLE => grouping
* JOINED => grouping, except when not targeting all concrete subtypes of the common supertype

Considering the only way of ending up in a situation where one type is being mass indexed but not its concrete subtype, is when that subtype is not indexed (`@Indexed(enabled=false)`), I think an acceptable approximation would be:

* TABLE_PER_CLASS => no grouping
* SINGLE_TABLE => grouping
* JOINED => grouping, except when one of the types involved has non-indexed concrete subtypes

I think the information "an indexed type has non-indexed concrete subtypes" could easily be inferred from current metadata in the constructor of `org.hibernate.search.mapper.pojo.mapping.impl.PojoTypeManagerContainer`: you can list subtypes using `org.hibernate.search.mapper.pojo.automaticindexing.building.impl.PojoImplicitReindexingResolverBuildingHelper#getConcreteEntitySubTypesForEntitySuperType` and check whether they are indexed using `org.hibernate.search.mapper.pojo.mapping.impl.PojoTypeManagerContainer.Builder#indexed` (change the map key type to `PojoRawTypeIdentifier` if that helps).
Checking whether that type is concrete is... left as an exercise to the reader ;D Not sure the info is available right there, but it's in `org.hibernate.search.engine.mapper.model.spi.MappableTypeModel#isAbstract` and in a couple other places.

From there you can pass the info to `org.hibernate.search.mapper.pojo.mapping.impl.AbstractPojoTypeManager.Builder#build` and expose it in `PojoLoadingTypeContext#hasNonIndexedConcreteSubtypes()`.
You'll be able to leverage it to prevent grouping in `org.hibernate.search.mapper.orm.loading.impl.AbstractHibernateOrmLoadingStrategy#groupingAllowed`.